# **2. Exploring data**

---

## **2.1 Prerequisites**



**Import libraries**

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

**Loading the dataset**

In [ ]:
df = pd.read_csv('lichess_games.csv')
df.head(5)

****

## **2.2 General tasks**

**How many rows and how many columns?**

In [8]:
rows, cols = df.shape
print(f'The dataset has {rows} rows and {cols} columns')

The dataset has 20058 rows and 16 columns


**What is the meaning of each row?**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20058 entries, 0 to 20057
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              20058 non-null  object 
 1   rated           20058 non-null  bool   
 2   created_at      20058 non-null  float64
 3   last_move_at    20058 non-null  float64
 4   turns           20058 non-null  int64  
 5   victory_status  20058 non-null  object 
 6   winner          20058 non-null  object 
 7   increment_code  20058 non-null  object 
 8   white_id        20058 non-null  object 
 9   white_rating    20058 non-null  int64  
 10  black_id        20058 non-null  object 
 11  black_rating    20058 non-null  int64  
 12  moves           20058 non-null  object 
 13  opening_eco     20058 non-null  object 
 14  opening_name    20058 non-null  object 
 15  opening_ply     20058 non-null  int64  
dtypes: bool(1), float64(2), int64(4), object(9)
memory usage: 2.3+ MB


In [12]:
df.head(1)

id  rated    created_at  last_move_at  turns victory_status winner  \
0  TZJHLljE  False  1.504210e+12  1.504210e+12     13      outoftime  white   

  increment_code  white_id  white_rating black_id  black_rating  \
0           15+2  bourgris          1500     a-00          1191   

                                               moves opening_eco  \
0  d4 d5 c4 c6 cxd5 e6 dxe6 fxe6 Nf3 Bb4+ Nc3 Ba5...         D10   

                       opening_name  opening_ply  
0  Slav Defense: Exchange Variation            5

Each row in the dataset represents **a single chess game**. It contains detailed information about that game, such as its unique ID, players, moves, ratings, game result, and more

**Are there duplicated rows?**

In [14]:
print(f"Nums of duplicated row: {df.duplicated().sum()}")

Nums of duplicated row: 429


There are 429 duplicated rows, therefore we have to remove them for further preprocessing.

In [15]:
df.drop_duplicates(inplace=True)

**What is the meaning of each column?**

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19629 entries, 0 to 20057
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              19629 non-null  object 
 1   rated           19629 non-null  bool   
 2   created_at      19629 non-null  float64
 3   last_move_at    19629 non-null  float64
 4   turns           19629 non-null  int64  
 5   victory_status  19629 non-null  object 
 6   winner          19629 non-null  object 
 7   increment_code  19629 non-null  object 
 8   white_id        19629 non-null  object 
 9   white_rating    19629 non-null  int64  
 10  black_id        19629 non-null  object 
 11  black_rating    19629 non-null  int64  
 12  moves           19629 non-null  object 
 13  opening_eco     19629 non-null  object 
 14  opening_name    19629 non-null  object 
 15  opening_ply     19629 non-null  int64  
dtypes: bool(1), float64(2), int64(4), object(9)
memory usage: 2.4+ MB


- **id**: A unique identifier assigned to each game.
- **rated**: The game rate (TRUE, FALSE).
- **created_at**: The game's start time (milliseconds).
- **last_move_at**: The timestamp of the game’s final move (milliseconds).
- **turns**: The total number of moves.
- **victory_status**: How the game ended ("outoftime," "resign," "mate").
- **winner**: Who won the game ("white", "black").
- **increment_code**: The time control format, (eg. "14+7," meaning 14 minutes plus a 7-second increment per move).
- **white_id**: white player's unique identifier.
- **white_rating**: The chess rating of the white player.
- **black_id**: black player's unique identifier.
- **black_rating**: The chess rating of the black player.
- **moves**: A list of all moves made during the game.
- **opening_eco**: Encyclopedia of Chess Openings code for the opening played.
- **opening_name**: The name of the chess opening used in the game.
- **opening_ply**: The number of half-moves (plies) played during the opening phase.

In [20]:
df.describe(include='all')

id  rated    created_at  last_move_at         turns  \
count      19629  19629  1.962900e+04  1.962900e+04  19629.000000   
unique     19113      2           NaN           NaN           NaN   
top     oCvN8zWL   True           NaN           NaN           NaN   
freq           2  15817           NaN           NaN           NaN   
mean         NaN    NaN  1.483721e+12  1.483722e+12     60.458607   
std          NaN    NaN  2.809364e+10  2.809353e+10     33.502852   
min          NaN    NaN  1.376772e+12  1.376772e+12      1.000000   
25%          NaN    NaN  1.477530e+12  1.477530e+12     37.000000   
50%          NaN    NaN  1.495830e+12  1.495830e+12     55.000000   
75%          NaN    NaN  1.503107e+12  1.503110e+12     79.000000   
max          NaN    NaN  1.504493e+12  1.504494e+12    349.000000   

       victory_status winner increment_code white_id  white_rating  black_id  \
count           19629  19629          19629    19629  19629.000000     19629   
unique              4      3            400     9438           NaN      9331   
top            resign  white           10+0     ssf7           NaN  king5891   
freq            10926   9792           7563       48           NaN        44   
mean              NaN    NaN            NaN      NaN   1596.090224       NaN   
std               NaN    NaN            NaN      NaN    290.165038       NaN   
min               NaN    NaN            NaN      NaN    784.000000       NaN   
25%               NaN    NaN            NaN      NaN   1400.000000       NaN   
50%               NaN    NaN            NaN      NaN   1567.000000       NaN   
75%               NaN    NaN            NaN      NaN   1791.000000       NaN   
max               NaN    NaN            NaN      NaN   2700.000000       NaN   

        black_rating  moves opening_eco          opening_name   opening_ply  
count   19629.000000  19629       19629                 19629  19629.000000  
unique           NaN  18920         365                  1477           NaN  
top              NaN  e4 e5         A00  Van't Kruijs Opening           NaN  
freq             NaN     27         992                   363           NaN  
mean     1588.324622    NaN         NaN                   NaN      4.804422  
std       290.164447    NaN         NaN                   NaN      2.792791  
min       789.000000    NaN         NaN                   NaN      1.000000  
25%      1393.000000    NaN         NaN                   NaN      3.000000  
50%      1562.000000    NaN         NaN                   NaN      4.000000  
75%      1783.000000    NaN         NaN                   NaN      6.000000  
max      2723.000000    NaN         NaN                   NaN     28.000000

Moreover, we can see that there are duplicated **id** values. Since **id** indicates unique identifier for each game, we consider rows having the same **id** is duplicated games.

In [22]:
print(f"Nums of total id: {df['id'].count()}")
print(f"Nums of distinct id: {df['id'].nunique()}")

Nums of total id: 19629
Nums of distinct id: 19113


In [24]:
df.drop_duplicates("id",keep="first",inplace=True)

In [25]:
df.describe(include='all')

id  rated    created_at  last_move_at         turns  \
count      19113  19113  1.911300e+04  1.911300e+04  19113.000000   
unique     19113      2           NaN           NaN           NaN   
top     N8G2JHGG   True           NaN           NaN           NaN   
freq           1  15467           NaN           NaN           NaN   
mean         NaN    NaN  1.483413e+12  1.483414e+12     60.513839   
std          NaN    NaN  2.833975e+10  2.833964e+10     33.488264   
min          NaN    NaN  1.376772e+12  1.376772e+12      1.000000   
25%          NaN    NaN  1.476910e+12  1.476912e+12     37.000000   
50%          NaN    NaN  1.495670e+12  1.495670e+12     55.000000   
75%          NaN    NaN  1.503080e+12  1.503080e+12     79.000000   
max          NaN    NaN  1.504493e+12  1.504494e+12    349.000000   

       victory_status winner increment_code white_id  white_rating black_id  \
count           19113  19113          19113    19113  19113.000000    19113   
unique              4      3            400     9438           NaN     9331   
top            resign  white           10+0     ssf7           NaN  docboss   
freq            10695   9545           7356       48           NaN       44   
mean              NaN    NaN            NaN      NaN   1597.300005      NaN   
std               NaN    NaN            NaN      NaN    290.022740      NaN   
min               NaN    NaN            NaN      NaN    784.000000      NaN   
25%               NaN    NaN            NaN      NaN   1401.000000      NaN   
50%               NaN    NaN            NaN      NaN   1567.000000      NaN   
75%               NaN    NaN            NaN      NaN   1792.000000      NaN   
max               NaN    NaN            NaN      NaN   2700.000000      NaN   

        black_rating  moves opening_eco      opening_name   opening_ply  
count   19113.000000  19113       19113             19113  19113.000000  
unique           NaN  18920         365              1477           NaN  
top              NaN  e4 e5         A00  Sicilian Defense           NaN  
freq             NaN     26         948               349           NaN  
mean     1590.045519    NaN         NaN               NaN      4.815780  
std       290.441694    NaN         NaN               NaN      2.798283  
min       789.000000    NaN         NaN               NaN      1.000000  
25%      1394.000000    NaN         NaN               NaN      3.000000  
50%      1563.000000    NaN         NaN               NaN      4.000000  
75%      1785.000000    NaN         NaN               NaN      6.000000  
max      2723.000000    NaN         NaN               NaN     28.000000

**What is the current data type of each column? Are there columns having inappropriate data types?**